### PR TITLE
Initialize Caffe2 only when running Caffe2 benchmarks

### PR DESCRIPTION
**Pull request by ilia-cher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20157 Support for Eigen thread pool
* #20087 Native ATen/Parallel backend
* #20057 Split ATen/Parallel into interface and backend
* #20050 Move inter-op settings into ATen/Parallel
* #20043 Port ATen/native to ATen/Parallel
* #20032 Port THNN to ATen/Parallel
* #20002 Remove explicit checks for parallelism from TH
* #19105 Port TH library to ATen/Parallel
* #19997 Intra-op parallel microbenchmarks for PT
* #19993 Support for non-contiguous tensors and arbitrary dtypes in PT benchmarks
* **#19980 Initialize Caffe2 only when running Caffe2 benchmarks**
* #19749 Allow a non-OpenMP based build

Summary:
At the moment Caffe2's GlobalInit conflicts with PyTorch
at::init_num_threads;

In this PR we initialize Caffe2 only when necessary - when Caffe2
benchmarks are requested (by default)

Test Plan:
BUILD_BINARY=1 BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop --cmake
cd benchmarks
python -m operator_benchmark.ops.matmul_test --framework PyTorch

Differential Revision: [D15229797](https://our.internmc.facebook.com/intern/diff/D15229797)

### CODE REVIEW

File: benchmarks/operator_benchmark/benchmark_core.py [hunk 1]
```diff
@@ -140,8 +140,10 @@ def run(self):
             # Currently, this is a sub-string matching.
             if self.args.operator and (self.args.operator not in full_test_id):
                 continue
-            if self.args.framework and (self.args.framework not in full_test_id):
-                continue
+            if self.args.framework:
+                frameworks = benchmark_utils.get_requested_frameworks(self.args.framework)
+                if all([fr not in full_test_id for fr in frameworks]):
```
Comment (dzhulgakov) [inline benchmarks/operator_benchmark/benchmark_core.py]: nit: [] is redundant

File: benchmarks/operator_benchmark/benchmark_runner.py [hunk 2]
```diff
@@ -73,13 +73,14 @@ def main():
 
     parser.add_argument(
         '--framework',
-        help='Run PyTorch or Caffe2 operators',
-        default=None)
+        help='Comma-delimited list of frameworks to test (Caffe2, PyTorch)',
+        default="Caffe2,PyTorch")
```
Comment (dzhulgakov) [inline benchmarks/operator_benchmark/benchmark_runner.py]: I think you can put `action=append` here and not rely on comma-separated list

**Review by dzhulgakov — APPROVED** [review]
minor suggestions, but looks good

**Comment by dzhulgakov** [comment]
And we should really unify initialization (also cc @smessmer). Would be nice to create an issue/quip with some plan about it

**Comment by ilia-cher** [comment]
Created https://github.com/pytorch/pytorch/issues/20189

File: benchmarks/operator_benchmark/benchmark_runner.py (hunk 2 above)
Comment (ilia-cher) [reply benchmarks/operator_benchmark/benchmark_runner.py]: not sure which one you mean, might be appendTuple, though comma-separated list of also pretty standard way to pass args

Comment (ilia-cher) [reply benchmarks/operator_benchmark/benchmark_runner.py]: though in this case we'll have to use `--framework` multiple times, I think comma-separated list is more standard and compact

**Pull request closed without merging**